# Data Cleaning Pipeline

In [410]:
# Import Libraries

import pandas as pd

In [411]:
# Edit Configuration 
print("Initial Output Rows:", pd.options.display.max_rows)
pd.options.display.max_rows = 60
print("Updated Output Rows:", pd.options.display.max_rows)

Initial Output Rows: 60
Updated Output Rows: 60


## Load Data

In [412]:
pump_data_val = pd.read_csv("data/training_set_values.csv")
pump_data_lab = pd.read_csv("data/training_set_labels.csv")

print("Pump Data Values Shape: \n")
print(pump_data_val.shape)

print("\nPump Data Labels Shape: \n")
print(pump_data_lab.shape)

Pump Data Values Shape: 

(59400, 40)

Pump Data Labels Shape: 

(59400, 2)


In [413]:
pump_data_val.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  ... payment_type  \
0  34.938093  -9.856322                  none            0  ...     annually   
1  34.698766  -2.147466              Zahanati            0  ...    never pay   
2  37.460664  -3.821329           Kwa Mahundi            0  ...   per bucket   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0  ...    never pay   
4  31.130847  -1.825359               Shuleni            0  ...    never pay   

  water_quality quality_group      quantity  quantity_group  \
0          soft          good        enough          enough   
1          soft          good  insufficient    insufficient   
2          soft          good        enough          enough   
3          soft          good           dry             dry   
4          soft          good      seasonal        seasonal   

                 source           source_type  source_class  \
0                spring                spring   groundwater   
1  rainwater harvesting  rainwater harvesting       surface   
2                   dam                   dam       surface   
3           machine dbh              borehole   groundwater   
4  rainwater harvesting  rainwater harvesting       surface   

               waterpoint_type waterpoint_type_group  
0           communal standpipe    communal standpipe  
1           communal standpipe    communal standpipe  
2  communal standpipe multiple    communal standpipe  
3  communal standpipe multiple    communal standpipe  
4           communal standpipe    communal standpipe  

[5 rows x 40 columns]

In [414]:
pump_data_lab.head()

id    status_group
0  69572      functional
1   8776      functional
2  34310      functional
3  67743  non functional
4  19728      functional

In [415]:
print("Check class Imbalance: ")
print(pump_data_lab.status_group.value_counts())

Check class Imbalance: 
status_group
functional                 32259
non functional             22824
functional needs repair     4317
Name: count, dtype: int64


## Data Cleaning

In [416]:
first_20_features = pump_data_val.iloc[:,:20]
first_20_features.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  \
0  34.938093  -9.856322                  none            0   
1  34.698766  -2.147466              Zahanati            0   
2  37.460664  -3.821329           Kwa Mahundi            0   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
4  31.130847  -1.825359               Shuleni            0   

                     basin  subvillage   region  region_code  district_code  \
0               Lake Nyasa    Mnyusi B   Iringa           11              5   
1            Lake Victoria     Nyamara     Mara           20              2   
2                  Pangani     Majengo  Manyara           21              4   
3  Ruvuma / Southern Coast  Mahakamani   Mtwara           90             63   
4            Lake Victoria  Kyanyamisa   Kagera           18              1   

         lga        ward  population public_meeting              recorded_by  
0     Ludewa    Mundindi         109           True  GeoData Consultants Ltd  
1  Serengeti       Natta         280            NaN  GeoData Consultants Ltd  
2  Simanjiro     Ngorika         250           True  GeoData Consultants Ltd  
3   Nanyumbu    Nanyumbu          58           True  GeoData Consultants Ltd  
4    Karagwe  Nyakasimbi           0           True  GeoData Consultants Ltd

In [417]:
first_20_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              59400 non-null  int64  
 1   amount_tsh      59400 non-null  float64
 2   date_recorded   59400 non-null  object 
 3   funder          55763 non-null  object 
 4   gps_height      59400 non-null  int64  
 5   installer       55745 non-null  object 
 6   longitude       59400 non-null  float64
 7   latitude        59400 non-null  float64
 8   wpt_name        59398 non-null  object 
 9   num_private     59400 non-null  int64  
 10  basin           59400 non-null  object 
 11  subvillage      59029 non-null  object 
 12  region          59400 non-null  object 
 13  region_code     59400 non-null  int64  
 14  district_code   59400 non-null  int64  
 15  lga             59400 non-null  object 
 16  ward            59400 non-null  object 
 17  population      59400 non-null 

### Datatype Analysis

In [418]:
# Convert to best possible datatype
first_20_features = first_20_features.convert_dtypes()
first_20_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              59400 non-null  Int64  
 1   amount_tsh      59400 non-null  Float64
 2   date_recorded   59400 non-null  string 
 3   funder          55763 non-null  string 
 4   gps_height      59400 non-null  Int64  
 5   installer       55745 non-null  string 
 6   longitude       59400 non-null  Float64
 7   latitude        59400 non-null  Float64
 8   wpt_name        59398 non-null  string 
 9   num_private     59400 non-null  Int64  
 10  basin           59400 non-null  string 
 11  subvillage      59029 non-null  string 
 12  region          59400 non-null  string 
 13  region_code     59400 non-null  Int64  
 14  district_code   59400 non-null  Int64  
 15  lga             59400 non-null  string 
 16  ward            59400 non-null  string 
 17  population      59400 non-null 

All the datatypes seem correct from domain knowledge exccept `date_recorded`. We can convert it manually.

In [419]:
first_20_features['date_recorded'] = first_20_features['date_recorded'].astype('datetime64[ns]')

In [420]:
first_20_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   id              59400 non-null  Int64         
 1   amount_tsh      59400 non-null  Float64       
 2   date_recorded   59400 non-null  datetime64[ns]
 3   funder          55763 non-null  string        
 4   gps_height      59400 non-null  Int64         
 5   installer       55745 non-null  string        
 6   longitude       59400 non-null  Float64       
 7   latitude        59400 non-null  Float64       
 8   wpt_name        59398 non-null  string        
 9   num_private     59400 non-null  Int64         
 10  basin           59400 non-null  string        
 11  subvillage      59029 non-null  string        
 12  region          59400 non-null  string        
 13  region_code     59400 non-null  Int64         
 14  district_code   59400 non-null  Int64         
 15  lg

### Uniform String Data across DataFrame

In [421]:
# Lowercase all the data and remove trailing whitespaces
string_cols = first_20_features.select_dtypes('string').columns
first_20_features[string_cols] = first_20_features[string_cols].apply(lambda x: x.str.lower().str.strip())

### Check for Zero-Variance Column

The columns with a single value do not add any purpose to the modeling, so let's check and remove it.

In [422]:
count_cols = first_20_features.nunique()
zer_var_cols = []
for col, count in count_cols.items():
    if count == 1:
        zer_var_cols.append(col)

print(f"Zero-Variance Columns: {zer_var_cols}")

Zero-Variance Columns: ['recorded_by']


As we can see the column - `recorded_by` contains only a single value, let's verify this and drop it from column list.

In [423]:
# Optional Step
first_20_features.recorded_by.value_counts()

recorded_by
geodata consultants ltd    59400
Name: count, dtype: Int64

In [424]:
cleaned_df = first_20_features.drop(zer_var_cols, axis=1)

### Check Duplicate Rows

In [425]:
cleaned_df.duplicated().sum()

0

No duplicate rows in the dataset

### Missing Value Analysis

In [426]:
# Check for null values
null_count = cleaned_df.isna().sum()
for col, count in null_count.items():
    if count > 0:
        print(f"Column: {col} contains {count} null values")


Column: funder contains 3637 null values
Column: installer contains 3655 null values
Column: wpt_name contains 2 null values
Column: subvillage contains 371 null values
Column: public_meeting contains 3334 null values


#### Column Analysis - Funder and Installer

In [427]:
# Check who are the different entities that fund the pumps
cleaned_df["funder"].value_counts(dropna=False)

funder
government of tanzania     9084
<NA>                       3637
danida                     3114
hesawa                     2202
rwssp                      1374
                           ... 
ms-danish                     1
juma                          1
paulo sange                   1
magul                         1
internal drainage basin       1
Name: count, Length: 1897, dtype: Int64

In [428]:
# Check who are the different entities that fund the pumps
cleaned_df["installer"].value_counts(dropna=False)

installer
dwe                         17405
<NA>                         3655
government                   1891
hesawa                       1395
rwe                          1206
                            ...  
upendo group                    1
insititutiona                   1
tukware enterp                  1
wizara  ya maji                 1
bingo foundation germany        1
Name: count, Length: 1936, dtype: Int64

In [429]:
# Check the funder of pump with no installer
cleaned_df.loc[cleaned_df["installer"].isna()==True, ["installer", "funder"]].value_counts(dropna=False)

installer  funder                        
<NA>       <NA>                              3584
           world bank                          11
           world vision                         7
           hesawa                               5
           ministry of water                    4
           government of tanzania               4
           rural water supply and sanitat       4
           community                            3
           ggm                                  2
           wfp                                  2
           district council                     1
           aar                                  1
           abddwe                               1
           bhws                                 1
           bgssws                               1
           donor                                1
           jica                                 1
           maro kyariga                         1
           koica and tanzania government        1
        

Most of the pumps with no installer information have no information on the funder. Let's check whether the pumps with no installer information but having funder information have common installers. 

In [430]:
# Get the list of funders for pumps with no installation info as list
no_installer_with_funder_list = cleaned_df[(cleaned_df.installer.isna()==True) & (cleaned_df.funder.isna()!=True)]['funder'].unique()

# Get the dataframe with all rows from the previous funders list
specific_funders_data = cleaned_df[cleaned_df.funder.isin(no_installer_with_funder_list)]
specific_funders_data

# Check what are the unique values for investors these funders
print(specific_funders_data[["funder", "installer"]].value_counts(dropna=False))

# Build a table of funders and installer
# pd.crosstab(specific_funders_data.funder, specific_funders_data.installer)

funder                  installer  
government of tanzania  dwe            4256
                        government     1637
hesawa                  dwe            1296
rwssp                   dwe             914
hesawa                  hesawa          850
                                       ... 
community               atisan            1
                        comunity          1
                        dmdd              1
district council        humfreys co       1
                        mtuwasa           1
Name: count, Length: 549, dtype: int64


There are multiple installers for same funders, therefore it is difficult to impute values without more domain knowledge. But we also observe there is a `not known` value in funders and installers, therefore we can replace the null values in both these columns to `not known`.

In [431]:
cleaned_df[['funder', 'installer']] = cleaned_df[['funder', 'installer']].fillna('not known')

#### Column Analysis - wpt_name

In [432]:
# Check who the funders and installers are for the pumps with no wpt_name
cleaned_df.loc[cleaned_df["wpt_name"].isna()==True, ["funder", "installer", "wpt_name"]].value_counts(dropna=False)

# Check if there are pther wpt_name for the same combination of funder and installer
specific_wpt_data = cleaned_df[(cleaned_df.funder.isin(['lawatefuka water supply', 'wsdp'])) | (cleaned_df.installer.isin(['lawatefuka water sup', 'aquarman drillers']))]
# specific_wpt_data[["funder", "installer", "wpt_name"]].value_counts(dropna=False)

# Investigate rows with null values
specific_wpt_data[specific_wpt_data.wpt_name.isna()]

id  amount_tsh date_recorded                   funder  gps_height  \
14595  31756         0.0    2012-10-22                     wsdp           0   
42935  16208         0.0    2013-03-15  lawatefuka water supply        1540   

                  installer  longitude  latitude wpt_name  num_private  \
14595     aquarman drillers  34.107893 -3.417657     <NA>            0   
42935  lawatefuka water sup  37.110999 -3.158049     <NA>            0   

          basin subvillage       region  region_code  district_code    lga  \
14595  internal    majengo    shinyanga           17              6  meatu   
42935   pangani      maini  kilimanjaro            3              7   siha   

                 ward  population  public_meeting  
14595          itinje           0            True  
42935  siha kaskazini           1            True

No pattern in `wpt_name`, therefore any sort of imputation might introduce bias in the data. Also, considering only 2 row have missing data, we can drop these rows.

In [433]:
cleaned_df = cleaned_df.dropna(subset=['wpt_name'])

In [434]:
cleaned_df.shape

(59398, 19)

#### Column Analysis - Subvillage

In [435]:
cleaned_df

id  amount_tsh date_recorded           funder  gps_height  \
0      69572      6000.0    2011-03-14            roman        1390   
1       8776         0.0    2013-03-06          grumeti        1399   
2      34310        25.0    2013-02-25     lottery club         686   
3      67743         0.0    2013-01-28           unicef         263   
4      19728         0.0    2011-07-13      action in a           0   
...      ...         ...           ...              ...         ...   
59395  60739        10.0    2013-05-03  germany republi        1210   
59396  27263      4700.0    2011-05-07      cefa-njombe        1212   
59397  37057         0.0    2011-04-11        not known           0   
59398  31282         0.0    2011-03-08            malec           0   
59399  26348         0.0    2011-03-23       world bank         191   

          installer  longitude   latitude              wpt_name  num_private  \
0             roman  34.938093  -9.856322                  none            0   
1           grumeti  34.698766  -2.147466              zahanati            0   
2      world vision  37.460664  -3.821329           kwa mahundi            0   
3            unicef  38.486161 -11.155298  zahanati ya nanyumbu            0   
4           artisan  31.130847  -1.825359               shuleni            0   
...             ...        ...        ...                   ...          ...   
59395           ces  37.169807  -3.253847   area three namba 27            0   
59396          cefa  35.249991  -9.070629     kwa yahona kuvala            0   
59397     not known  34.017087  -8.750434               mashine            0   
59398          musa  35.861315  -6.378573                mshoro            0   
59399         world  38.104048  -6.747464       kwa mzee lugawa            0   

                         basin    subvillage       region  region_code  \
0                   lake nyasa      mnyusi b       iringa           11   
1                lake victoria       nyamara         mara           20   
2                      pangani       majengo      manyara           21   
3      ruvuma / southern coast    mahakamani       mtwara           90   
4                lake victoria    kyanyamisa       kagera           18   
...                        ...           ...          ...          ...   
59395                  pangani      kiduruni  kilimanjaro            3   
59396                   rufiji      igumbilo       iringa           11   
59397                   rufiji     madungulu        mbeya           12   
59398                   rufiji        mwinyi       dodoma            1   
59399              wami / ruvu  kikatanyemba     morogoro            5   

       district_code             lga              ward  population  \
0                  5          ludewa          mundindi         109   
1                  2       serengeti             natta         280   
2                  4       simanjiro           ngorika         250   
3                 63        nanyumbu          nanyumbu          58   
4                  1         karagwe        nyakasimbi           0   
...              ...             ...               ...         ...   
59395              5             hai  masama magharibi         125   
59396              4          njombe            ikondo          56   
59397              7         mbarali           chimala           0   
59398              4        chamwino      mvumi makulu           0   
59399              2  morogoro rural        ngerengere         150   

       public_meeting  
0                True  
1                <NA>  
2                True  
3                True  
4                True  
...               ...  
59395            True  
59396            True  
59397            True  
59398            True  
59399            True  

[59398 rows x 19 columns]

In [436]:
# Analyse rows with null values in subvillage
filter_subvillage_null = cleaned_df[cleaned_df.subvillage.isna()]

In [437]:
filter_subvillage_null[['region', 'subvillage']].value_counts(dropna=False)

region  subvillage
dodoma  <NA>          361
mwanza  <NA>           10
Name: count, dtype: int64

In [438]:
cleaned_df[cleaned_df.subvillage.isin(['dodoma', 'mwanza'])].groupby(['subvillage'])['region'].value_counts()

subvillage  region   
dodoma      mtwara       17
            lindi        12
            ruvuma        9
            singida       5
            shinyanga     3
            kagera        2
            mara          2
            rukwa         2
mwanza      morogoro      4
            mwanza        3
            shinyanga     2
            mtwara        1
Name: count, dtype: int64

In [439]:
# Check for any pattern with other columns
cleaned_df[cleaned_df.subvillage.isin(['dodoma', 'mwanza',pd.NA])].groupby(['region', 'subvillage'], dropna=False)[['lga']].value_counts(dropna=False)

region     subvillage  lga             
dodoma     <NA>        kongwa              361
kagera     dodoma      chato                 2
lindi      dodoma      ruangwa              10
                       lindi rural           1
                       nachingwea            1
mara       dodoma      bunda                 2
morogoro   mwanza      ulanga                4
mtwara     dodoma      masasi                7
                       newala                5
                       tandahimba            2
                       mtwara rural          2
                       nanyumbu              1
           mwanza      masasi                1
mwanza     mwanza      geita                 3
           <NA>        ukerewe              10
rukwa      dodoma      sumbawanga rural      2
ruvuma     dodoma      songea rural          5
                       namtumbo              4
shinyanga  dodoma      kishapu               2
                       maswa                 1
           mwanza   

In [440]:
print(cleaned_df[(cleaned_df.region.isin(['dodoma', 'mwanza'])) & (cleaned_df.subvillage.isin(['dodoma', 'mwanza',pd.NA]))].groupby(['region', 'ward'])[["subvillage", "lga"]].value_counts(dropna=False))
dodoma_ward_list = cleaned_df[(cleaned_df.region=='dodoma') & (cleaned_df.subvillage.isin(['dodoma', 'mwanza',pd.NA]))]["ward"].unique()
mwanza_ward_list = cleaned_df[(cleaned_df.region=='mwanza') & (cleaned_df.subvillage.isin(['dodoma', 'mwanza',pd.NA]))]["ward"].unique()

region  ward          subvillage  lga    
dodoma  chamkoroma    <NA>        kongwa     41
        hogoro        <NA>        kongwa     25
        iduo          <NA>        kongwa     16
        kibaigwa      <NA>        kongwa      7
        kongwa urban  <NA>        kongwa     19
        mkoka         <NA>        kongwa     26
        mlali         <NA>        kongwa     41
        mtanana       <NA>        kongwa     19
        njoge         <NA>        kongwa     23
        pandambili    <NA>        kongwa     46
        sagara        <NA>        kongwa     22
        sejeli        <NA>        kongwa     34
        ugogoni       <NA>        kongwa     26
        zoissa        <NA>        kongwa     16
mwanza  bukanda       <NA>        ukerewe     1
        bukiko        <NA>        ukerewe     3
        bukindo       <NA>        ukerewe     1
        bwiro         <NA>        ukerewe     1
        bwisya        <NA>        ukerewe     2
        murutunguru   <NA>        ukerewe     

In [441]:
cleaned_df[cleaned_df.lga.isin(['kongwa', 'ukerewe'])].groupby(['lga'])[['subvillage']].value_counts(dropna=False)

lga      subvillage
kongwa   <NA>          361
ukerewe  kabakara       13
         <NA>           10
         center          9
         buyoga          7
                      ... 
         apembe          1
         azimio a        1
         umoja           1
         uwanjani        1
         zamu a          1
Name: count, Length: 185, dtype: int64

In [442]:
# cleaned_df[(cleaned_df.region=='dodoma') & (cleaned_df.subvillage.isin(['dodoma', 'mwanza',pd.NA]))]
# cleaned_df[(cleaned_df.region.isin(['dodoma'])) & (cleaned_df.ward.isin(dodoma_ward_list))].groupby(['region', 'ward'])[['subvillage']].value_counts(dropna=False)
cleaned_df[(cleaned_df.region.isin(['mwanza'])) & (cleaned_df.ward.isin(mwanza_ward_list))].groupby(['region', 'ward'])[['subvillage']].value_counts(dropna=False)

region  ward         subvillage
mwanza  bukanda      kwaya         5
                     hamkuno       3
                     ubungeni      2
                     mwaloni       2
                     mkuyuni       2
                                  ..
        nyang'hwale  n/center      1
                     isengwa       1
                     igalula       1
                     center        1
                     ikwendele     1
Name: count, Length: 113, dtype: int64

From this analysis, we see that `dodoma` and `mwanza` are valid subvillage values and the missing values have regions as same value for both these values. Therefore, we can impute the missing values in subvillage with region value.

> Note: We could have dropped these rows as we have sufficient data points but I plan to group categories with less number
of values togeter, so imputation might be useful.

In [443]:
cleaned_df['subvillage'] = cleaned_df['subvillage'].fillna(cleaned_df['region'])

In [444]:
cleaned_df.isna().sum()

id                   0
amount_tsh           0
date_recorded        0
funder               0
gps_height           0
installer            0
longitude            0
latitude             0
wpt_name             0
num_private          0
basin                0
subvillage           0
region               0
region_code          0
district_code        0
lga                  0
ward                 0
population           0
public_meeting    3334
dtype: int64

#### Column Analysis - public_meeting

In [445]:
cleaned_df.public_meeting.value_counts(dropna=False)

public_meeting
True     51009
False     5055
<NA>      3334
Name: count, dtype: Int64

In [446]:
cleaned_df[cleaned_df.public_meeting.isna()]

id  amount_tsh date_recorded        funder  gps_height  \
1       8776         0.0    2013-03-06       grumeti        1399   
18     34169         0.0    2011-07-22        hesawa        1162   
21      6091         0.0    2013-02-10          dwsp           0   
58     24593         0.0    2013-01-22          kkkt        1703   
73     35715         0.0    2013-02-27        hesawa        1565   
...      ...         ...           ...           ...         ...   
59306  71419         0.0    2013-01-16        cocern        1296   
59310  32815         0.0    2013-04-04         tasaf         892   
59324    748         0.0    2013-01-22  world vision           0   
59344   6450         0.0    2013-10-03     not known        1303   
59387  26640       100.0    2013-03-12             0          25   

          installer  longitude  latitude             wpt_name  num_private  \
1           grumeti  34.698766 -2.147466             zahanati            0   
18              dwe  32.920154 -1.947868               ngomee            0   
21              dwe        0.0      -0.0             muungano            0   
58             kkkt  35.561346 -3.806879  kwa iyora mgirigisi            0   
73           hesawa  34.657034  -1.88802             machumbe            0   
...             ...        ...       ...                  ...          ...   
59306         twesa  30.930385 -3.300592           kwa mhunda            0   
59310         tasaf   37.74409 -4.062296          ward office            0   
59324  world vision        0.0      -0.0             mwazwilo            0   
59344     not known  36.900911 -3.111477   pentecoste swidish            0   
59387             0   39.17648 -6.957098           kwa maliba            0   

                 basin          subvillage         region  region_code  \
1        lake victoria             nyamara           mara           20   
18       lake victoria              center         mwanza           19   
21       lake victoria          ibabachegu      shinyanga           17   
58            internal              ants b        manyara           21   
73       lake victoria             mtakuja           mara           20   
...                ...                 ...            ...          ...   
59306  lake tanganyika        nguvu kazi a         kigoma           16   
59310          pangani                same    kilimanjaro            3   
59324    lake victoria               mbita      shinyanga           17   
59344         internal            madukani         arusha            2   
59387      wami / ruvu  mtaa wa yangeyange  dar es salaam            7   

       district_code        lga               ward  population  public_meeting  
1                  2  serengeti              natta         280            <NA>  
18                 1    ukerewe           ilangala        1000            <NA>  
21                 1    bariadi  ikungulyabashashi           0            <NA>  
58                 3      mbulu            bargish         456            <NA>  
73                 2  serengeti           manchira         200            <NA>  
...              ...        ...                ...         ...             ...  
59306              1    kibondo            kakonko           1            <NA>  
59310              3       same         same urban          65            <NA>  
59324              1    bariadi              mbita           0            <NA>  
59344              7       meru       ngarenanyuki         230            <NA>  
59387              2      ilala           msongola         100            <NA>  

[3334 rows x 19 columns]

In [447]:
# Analyze null rows
cleaned_df[cleaned_df.public_meeting.isna()][["district_code"]].value_counts(dropna=False)

district_code
3                1015
2                1012
1                 486
7                 288
5                 141
4                 108
13                 79
43                 73
6                  59
53                 28
33                 23
8                  15
63                  2
23                  2
80                  2
30                  1
Name: count, dtype: int64

We don't have enough information to impute the `public_meeting` column. Therefore we will drop the rows with null values.

In [448]:
cleaned_df = cleaned_df.dropna(subset=['public_meeting'])

In [449]:
cleaned_df.shape

(56064, 19)

In [450]:
cleaned_df.isna().sum()

id                0
amount_tsh        0
date_recorded     0
funder            0
gps_height        0
installer         0
longitude         0
latitude          0
wpt_name          0
num_private       0
basin             0
subvillage        0
region            0
region_code       0
district_code     0
lga               0
ward              0
population        0
public_meeting    0
dtype: int64

### Categorical data cleaning

In [451]:
combined_cols = []
cleaned_df.shape

(56064, 19)

In [452]:
# Print all categorical columns
cleaned_df.select_dtypes('string').columns

Index(['funder', 'installer', 'wpt_name', 'basin', 'subvillage', 'region',
       'lga', 'ward'],
      dtype='object')

In [453]:
#Helper Function
def combine_categories(data, column, thresh=5):
    series_count = pd.value_counts(data[column])
    mask = ((series_count / series_count.sum())*100).lt(thresh)
    grouped_cat = series_count[mask].index.tolist()
    data = data.assign(**{f'{column}_combined':data[column].apply(lambda x: 'Other' if x in grouped_cat else x).values})
    if column not in combined_cols:
        combined_cols.append(column)
    print(f"Combined Columns List: {combined_cols}")
    return data

#### Column Analysis - Funder

In [454]:
cleaned_df.funder.value_counts()

funder
government of tanzania    8666
not known                 3466
danida                    2924
hesawa                    1986
kkkt                      1282
                          ... 
kaaya                        1
kalitesi                     1
sun-ja na                    1
twice                        1
rwi                          1
Name: count, Length: 1754, dtype: Int64

We have `1754` unique values in column funder with almost 18% values in one category. We can assign the funder with very low frequency to be `Other` as this will help reduce the complexity of model without loosing too much information. We will combine all the categories with frequency less than 1%.

In [455]:
cleaned_df = combine_categories(cleaned_df, 'funder', 1)
cleaned_df.funder_combined.value_counts()

Combined Columns List: ['funder']


funder_combined
Other                     26636
government of tanzania     8666
not known                  3466
danida                     2924
hesawa                     1986
kkkt                       1282
world bank                 1273
rwssp                      1256
world vision               1172
unicef                     1031
dhv                         828
tasaf                       821
district council            818
private individual          792
norad                       760
germany republi             610
tcrs                        594
water                       583
ministry of water           566
Name: count, dtype: int64

#### Column Analysis - Installer

In [456]:
cleaned_df.installer.value_counts()

installer
dwe                     16683
not known                3484
government               1541
rwe                      1190
hesawa                   1178
                        ...  
halimashauli                1
rudri                       1
alex moyela                 1
centra government           1
compasion internatio        1
Name: count, Length: 1779, dtype: Int64

Similarly, we have too many unique values with low frequencies. Therefore, we will combine them

In [457]:
cleaned_df = combine_categories(cleaned_df, 'installer', 1)
cleaned_df.installer_combined.value_counts()

Combined Columns List: ['funder', 'installer']


installer_combined
Other                 24885
dwe                   16683
not known              3484
government             1541
rwe                    1190
hesawa                 1178
commu                  1064
district council        950
kkkt                    905
danida                  863
central government      757
tcrs                    697
world vision            633
danid                   624
ces                     610
Name: count, dtype: int64

#### Column Analysis - wpt_name

In [458]:
cleaned_df.wpt_name.value_counts()

wpt_name
none                 3517
shuleni              1602
zahanati              765
msikitini             520
kanisani              313
                     ... 
riata                   1
kwa eliot masawa        1
kwa james siai          1
kwa doritha             1
shuleni kwalakama       1
Name: count, Length: 35497, dtype: Int64

Similarly,we combine categories with frequency less than 1%

In [459]:
cleaned_df = combine_categories(cleaned_df, 'wpt_name', 1)
cleaned_df.wpt_name_combined.value_counts()

Combined Columns List: ['funder', 'installer', 'wpt_name']


wpt_name_combined
Other       50180
none         3517
shuleni      1602
zahanati      765
Name: count, dtype: int64

#### Column Analysis - Basin

In [460]:
cleaned_df.basin.value_counts()

basin
pangani                    8902
lake victoria              8856
rufiji                     7949
internal                   7465
lake tanganyika            6143
wami / ruvu                5506
lake nyasa                 4694
ruvuma / southern coast    4137
lake rukwa                 2412
Name: count, dtype: Int64

No transformation needed.

#### Column Analysis - Subvillage

In [461]:
cleaned_df.subvillage.value_counts()

subvillage
shuleni         492
majengo         490
madukani        453
dodoma          409
kati            359
               ... 
kwamvuma          1
malesela          1
miembe dekwa      1
lung'uda          1
mbanja b          1
Name: count, Length: 18367, dtype: Int64

We have a lot of categories with low frequencies. Therefore we will combine categories with less than 0.5% frequency.

In [462]:
cleaned_df = combine_categories(cleaned_df, 'subvillage', 0.5)
cleaned_df.subvillage_combined.value_counts()

Combined Columns List: ['funder', 'installer', 'wpt_name', 'subvillage']


subvillage_combined
Other       53861
shuleni       492
majengo       490
madukani      453
dodoma        409
kati          359
Name: count, dtype: int64

#### Column Analysis - Region

In [463]:
cleaned_df.region.value_counts()

region
iringa           5292
mbeya            4617
kilimanjaro      4358
shinyanga        4335
morogoro         3974
kagera           3316
arusha           3312
mwanza           2814
kigoma           2785
pwani            2628
tanga            2544
dodoma           2200
ruvuma           2118
singida          1968
tabora           1902
rukwa            1755
mtwara           1698
manyara          1537
lindi            1365
mara             1201
dar es salaam     345
Name: count, dtype: Int64

In [464]:
cleaned_df = combine_categories(cleaned_df, 'region', 3)
cleaned_df.region_combined.value_counts()

Combined Columns List: ['funder', 'installer', 'wpt_name', 'subvillage', 'region']


region_combined
iringa         5292
mbeya          4617
Other          4448
kilimanjaro    4358
shinyanga      4335
morogoro       3974
kagera         3316
arusha         3312
mwanza         2814
kigoma         2785
pwani          2628
tanga          2544
dodoma         2200
ruvuma         2118
singida        1968
tabora         1902
rukwa          1755
mtwara         1698
Name: count, dtype: int64

#### Column Analysis - LGA

In [465]:
cleaned_df.lga.value_counts()

lga
njombe          2503
arusha rural    1248
moshi rural     1234
rungwe          1089
kilosa          1070
                ... 
kigoma urban      69
arusha urban      63
ilala             37
lindi urban       21
nyamagana          1
Name: count, Length: 125, dtype: Int64

In [466]:
cleaned_df = combine_categories(cleaned_df, 'lga', 1)
cleaned_df.lga_combined.value_counts()

Combined Columns List: ['funder', 'installer', 'wpt_name', 'subvillage', 'region', 'lga']


lga_combined
Other              25436
njombe              2503
arusha rural        1248
moshi rural         1234
rungwe              1089
kilosa              1070
kasulu              1045
mbozi               1029
bagamoyo             993
meru                 990
kilombero            956
singida rural        880
same                 874
bariadi              871
kyela                859
kibondo              853
kigoma rural         818
maswa                800
karagwe              771
magu                 769
iringa rural         728
kahama               720
lushoto              694
songea rural         692
mvomero              671
namtumbo             671
ngara                669
mpanda               669
ulanga               660
makete               629
mbarali              626
hai                  625
kwimba               607
rombo                594
shinyanga rural      585
nzega                573
ludewa               563
Name: count, dtype: int64

#### Column Analysis - Ward

In [467]:
cleaned_df.ward.value_counts()

ward
igosi           307
imalinyi        252
siha kati       232
mdandu          231
nduruma         216
               ... 
chinugulu         1
kinungu           1
ifinga            1
thawi             1
mbinga urban      1
Name: count, Length: 2049, dtype: Int64

In [468]:
cleaned_df = combine_categories(cleaned_df, 'ward', 0.4)
cleaned_df.ward_combined.value_counts()

Combined Columns List: ['funder', 'installer', 'wpt_name', 'subvillage', 'region', 'lga', 'ward']


ward_combined
Other        55042
igosi          307
imalinyi       252
siha kati      232
mdandu         231
Name: count, dtype: int64

#### Drop altered columns

In [406]:
cleaned_first_20_df = cleaned_df.drop(combined_cols, axis=1)
cleaned_first_20_df = cleaned_first_20_df.convert_dtypes()

In [407]:
cleaned_first_20_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 56064 entries, 0 to 59399
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   id                   56064 non-null  Int64         
 1   amount_tsh           56064 non-null  Float64       
 2   date_recorded        56064 non-null  datetime64[ns]
 3   gps_height           56064 non-null  Int64         
 4   longitude            56064 non-null  Float64       
 5   latitude             56064 non-null  Float64       
 6   num_private          56064 non-null  Int64         
 7   basin                56064 non-null  string        
 8   region_code          56064 non-null  Int64         
 9   district_code        56064 non-null  Int64         
 10  population           56064 non-null  Int64         
 11  public_meeting       56064 non-null  boolean       
 12  funder_combined      56064 non-null  string        
 13  installer_combined   56064 non-null 

#### Numerical Column Analysis

In [489]:
drop_numerical_cols = []
list(set(cleaned_df.columns).difference(set(cleaned_df.select_dtypes(['string','object']))))

['id',
 'district_code',
 'region_code',
 'public_meeting',
 'gps_height',
 'date_recorded',
 'longitude',
 'population',
 'latitude',
 'num_private',
 'amount_tsh']

We can drop id as it has all unique elements

In [490]:
for col in cleaned_df.columns:
    print(f"{col} : {len(cleaned_df[col].unique())}")

id : 56064
amount_tsh : 94
date_recorded : 354
funder : 1754
gps_height : 2428
installer : 1779
longitude : 54564
latitude : 54565
wpt_name : 35497
num_private : 62
basin : 9
subvillage : 18367
region : 21
region_code : 27
district_code : 20
lga : 125
ward : 2049
population : 1026
public_meeting : 2
funder_combined : 19
installer_combined : 15
wpt_name_combined : 4
subvillage_combined : 6
region_combined : 18
lga_combined : 37
ward_combined : 5


#### Column Analysis - District Code

In [476]:
cleaned_df[['district_code', 'region_code']].value_counts()

district_code  region_code
4              11             2473
3              17             1282
4              3              1208
2              2              1202
               16             1156
                              ... 
6              6                26
0              1                23
80             17               10
67             60                6
43             40                1
Name: count, Length: 130, dtype: int64

#### Column Analysis - Region Code

#### Column Analysis - Public Meeting

#### Column Analysis - GPS Height

#### Column Analysis - Date Recorded

#### Column Analysis - Population

#### Column Analysis - Num Private

#### Column Analysis - Amount TSH